src/protocol/BaseProverManager.sol

```diff
@@ -0,0 +1,383 @@
+        uint40 end;
+        // the time by which the prover needs to submit a proof
+        uint40 deadline;
+        // whether the proof came after the deadline
+        bool pastDeadline;
+    }
+
+    address public immutable inbox;
+    ICheckpointTracker public immutable checkpointTracker;
+    IPublicationFeed public immutable publicationFeed;
+
+    /// @notice Common balances for proposers and provers
+    mapping(address user => uint256 balance) private _balances;
+    /// @notice The current period
+    uint256 private _currentPeriodId;
+    /// @dev Periods represent proving windows
+    mapping(uint256 periodId => Period) private _periods;
+
+    /// @dev Initializes the contract state and deposits the initial prover's liveness bond.
+    /// The constructor also calls `_claimProvingVacancy`. Publications will actually start in period 1.
+    /// @param _inbox The address of the inbox contract
+    /// @param _checkpointTracker The address of the checkpoint tracker contract
+    /// @param _publicationFeed The address of the publication feed contract
+    /// @param _initialProver The address that will be designated as the initial prover
+    /// @param _initialFee The fee for the initial period
+    /// @param _initialDeposit The initial deposit that will be added to the `_initialProver`'s balance
+    constructor(
+        address _inbox,
+        address _checkpointTracker,
+        address _publicationFeed,
+        address _initialProver,
+        uint96 _initialFee,
+        uint256 _initialDeposit
+    ) {
+        require(_inbox != address(0), "Inbox address cannot be 0");
+        require(_checkpointTracker != address(0), "Checkpoint tracker address cannot be 0");
+        require(_publicationFeed != address(0), "Publication feed address cannot be 0");
+        require(_initialProver != address(0), "Initial prover address cannot be 0");
+
+        inbox = _inbox;
+        checkpointTracker = ICheckpointTracker(_checkpointTracker);
+        publicationFeed = IPublicationFeed(_publicationFeed);
+
+        // Close the first period so every period has a previous one (and an implicit start timestamp)
+        // The initial fee and prover will take effect in the block after this one
+        _deposit(_initialProver, _initialDeposit);
+        _claimProvingVacancy(_initialFee, _initialProver);
+    }
+
+    /// @notice Withdraw available(unlocked) funds.
+    /// @param amount The amount to withdraw
+    function withdraw(uint256 amount) external {
+        _balances[msg.sender] -= amount;
+        _transferOut(msg.sender, amount);
+        emit Withdrawal(msg.sender, amount);
+    }
+
+    /// @inheritdoc IProposerFees
+    /// @dev This function advances to the next period if the current period has ended.
+    function payPublicationFee(address proposer, bool isDelayed) external {
+        require(msg.sender == inbox, "Only the Inbox contract can call this function");
```

Comment:
There are several examples of error strings rather than custom errors (as required by the style guide). It's probably worth creating them as we go.